extensions/EXT/EGL_EXT_surface_compression.txt

```diff
@@ -0,0 +1,289 @@
+Name
+
+    EXT_surface_compression
+
+Name Strings
+
+    EGL_EXT_surface_compression
+
+Contributors
+
+    Jan-Harald Fredriksen, Arm
+    Lisa Wu, Arm
+    George Liu, Arm
+    Laurie Hedge, Imagination Technologies
+
+Contact
+
+    Jan-Harald Fredriksen (jan-harald.fredriksen 'at' arm.com)
+
+IP Status
+
+    No known IP claims.
+
+Status
+
+    Complete
+
+Version
+
+    Version 1 - November 15, 2021
+
+Number
+
+    EGL Extension #147
+
+Dependencies
+
+    These extensions are written against the wording of the EGL 1.5
+    specification (August 27, 2014).
+
+    This extension interacts with EGL_EXT_yuv_surface.
+
+Overview
+
+    Applications may wish to take advantage of framebuffer compression. Some
+    platforms may support framebuffer compression at fixed bitrates. Such
+    compression algorithms generally produce results that are visually lossless,
+    but the results are typically not bit exact when compared to a non-compressed
+    result.
+
+    This extension enables applications to opt-in to fixed-rate compression
+    for EGL window surfaces.
+
+    Compression may not be supported for all framebuffer formats. It can still
+    be requested for all formats and applications can query what level of compression
+    was actually enabled. 
+
+New Procedures and Functions
+
+    EGLBoolean eglQuerySupportedCompressionRatesEXT(
+           EGLDisplay dpy, EGLConfig config, const EGLAttrib *attrib_list,
+           EGLint *rates, EGLint rate_size, EGLint *num_rates);
+
+New Tokens
+
+    New attributes accepted by the <attrib_list> argument of
+    eglCreatePlatformWindowSurface and eglCreateWindowSurface:
+        EGL_SURFACE_COMPRESSION_EXT                     0x34B0
+
+    [Only if EGL_EXT_yuv_surface is supported]
+        EGL_SURFACE_COMPRESSION_PLANE1_EXT              0x328E
+        EGL_SURFACE_COMPRESSION_PLANE2_EXT              0x328F
+
+    Accepted as attribute values for EGL_SURFACE_COMPRESSION_EXT by
+    eglCreatePlatformWindowSurface and eglCreateWindowSurface:
+        EGL_SURFACE_COMPRESSION_FIXED_RATE_NONE_EXT     0x34B1
+        EGL_SURFACE_COMPRESSION_FIXED_RATE_DEFAULT_EXT  0x34B2
+
+        EGL_SURFACE_COMPRESSION_FIXED_RATE_1BPC_EXT     0x34B4

```

Comment:
This is very specific to me. What is the benefit to the application to know the compression rate?
Lets take for example ASTC - you get a a compression block, it tells you some information about format. Which might be something that is compatible with Window System or not. So that is my question rate vs format, seems format would be more useful. And would that then lead to having configs so you can choose a compression format? Or is this YUV specific compression on top of a YUV format? How is that rate passed to windows system? decoded from buffer?

Reply:
There are some long discussion on this in the Khronos internal issues. 
New formats was an option, but lead to a very large number of new formats - as most vendors support a small handful of compression rates per format.
How the rate is passed around is somewhat an implementation detail, but could be done with private Gralloc flags or DRM format modifiers.

Reply:
The buffer its self would be un compressed size then, And app will pick compression. 
It reminds me a bit of the old BAUD rate for dial up modems - lots of rates 9600 ->38k. But I think in use, wont an app either pick no compression, or maximum compression? I'm just not sure an app/developer/user would choose one over another with so many choices. How would that be selected or determined? Just curios. Or in practice, would you expect a limited set?

Reply:
Yes, in practice, each implementation will most likely support a subset of bitrates per format. For our implementation, it's 3 rates per format. eglQuerySupportedCompressionRatesEXT will return the rates the given implementation actually supports.
We do provide the EGL_SURFACE_COMPRESSION_FIXED_RATE_DEFAULT_EXT option that developers can use if they just want the implementation to pick a "best" compression rates. The specific rates are there for anyone wanting to fine-tune beyond that (and also that we can allow queries to return some information of what compression rate you actually got).

Reply:
This seems like it should all be covered by DRM format modifiers and application or vendor-specific logic layered on top of it as desired. Why is a further extension needed/useful, and how would the two interact?

Reply:
Our implementation will use this information - together with various internal state - to decide what DRM format modifiers to apply. The DRM format modifiers will be vendor specific. We'll likely make different decisions for different generations of hardware.
I can't speak to what other vendors plan to do.
In that sense, this API acts as an abstraction across vendors and HW generations. This also keeps the selection mechanism the same between this and the corresponding GLES extension, which also applies to driver-internal images.

Reply:
Thanks